Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,107 @@
+
+The current hosting solutions are defined at [Documentation/design-docs/host-components.md](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md). Along with the existing hosts an additional COM activation host library will be added. This library will export the required functions for COM class activation.
+
+>[`HRESULT DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID *ppv);`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllgetclassobject)
+
+>[`HRESULT DllCanUnloadNow();`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllcanunloadnow)
+
+When `DllGetClassObject()` is called in an activation scenario, the following will occur:
+
+1) Determine additional registration information needed for activation.
+    * **The location and mechanics of accessing this registration information is still TBD**
+    * `Assembly` (**required**) - the [Fully-Qualified Name](https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) of the assembly.
+    * `Class` (**required**) - the full type name (e.g. `MyCompany.MyProduct.MyClass`).
+    * `Codebase` (**required**) - an absolute path to the assembly to load. If a [`runtimeconfig.json`](https://github.com/dotnet/cli/blob/master/Documentation/specs/runtime-configuration-file.md) file exists adjacent to the assembly that file will be used to describe runtime configuration details.
+    * `RuntimeVersion` (optional) - if not supplied the latest CLR found will be used.
+1) Using the existing `hostfxr` library, attempt to discover the desired CLR.
+    * A current running CLR will be preferred, otherwise a new CLR instance will be created.
+1) A request to the CLR will be made via a new method for class activation within a COM environment.
+    * The ability to load the assembly and then create a class instance will require a function be exposed that can be called from `hostfxr`.
+    * Example of API in `System.Private.CoreLib` on the `Activator` class:
+        ``` csharp
+        public static class Activator
+        {
+            ...
+            #if WINDOWS
+            public static ObjectHandle CreateInstanceForCom(string assemblyName, string typeName);
+            #endif
+            ...
+        }
+        ```
+        Note this API would not be exposed outside of `System.Private.CoreLib`.
+    * The loading of the assembly will take place in a new [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.1) for dependency isolation.
+1) The class instance will be returned to the caller of `DllGetClassObject()`.
+
+In order to properly support `DllCanUnloadNow()`, .NET Core will be required to expose an API that permits the querying of how many COM activated classes are currently outstanding.
```

Comment:
Do we support unload managed assembly in dotnetcore?

Reply:
Good question. I think if we are using the ALC that is possible, but I don't know for sure.

Reply:
We might, but this very hard to get right. .NET Framework just returns false from `DllCanUnloadNow`. I think we should do the same.

Reply:
Okay, if there is no appetite for this support, we will document it as always returning false until needed.